### PR TITLE
Device index patch

### LINKED ISSUE #1748: Build failure due to `c10::DeviceIndex`.
https://gitlab-master.nvidia.com/dl/pytorch/update-scripts/-/jobs/82329135

See also the Nightly-CI teams chat.

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #1748 

guard c10::cuda::GetDevice API change on TORCH_VERSION

with this change, it ensures that we can build against stable release `< 2.2.0`, as well as TOT after  https://github.com/pytorch/pytorch/pull/119142

For 2.3.0 nightly, if someone accidentally checkout a commit before the patch, the build will still fail.

### CODE REVIEW

**Comment by Priya2698** [comment]
LGTM. Missed the torch version requirement in the earlier patch. 
Thanks for fixing.

**Review by Priya2698 — APPROVED** [review]
LGTM. Missed the torch version requirement in the earlier patch.
Thanks for fixing.

**Comment by jjsjann123** [comment]
cc'ing @tfogal @xwang233 , wondering if there's a better practice when we chase nightly ToT.

**Comment by jjsjann123** [comment]
!build

**Comment by xwang233** [comment]
There seems to be a build failure on distributed module happening at the same time. 😟

**Comment by naoyam** [comment]
@samnordmann @cowanmeg

**Comment by samnordmann** [comment]
I have check out to this branch but am still getting the error
```
FAILED: CMakeFiles/nvfuser_tests.dir/test/test_gpu_compute_with.cpp.o
/usr/bin/c++ -DNVFUSER_DISTRIBUTED -DUSE_C10D_GLOO -DUSE_C10D_MPI -DUSE_C10D_NCCL -DUSE_C10D_UCC -DUSE_DISTRIBUTED -DUSE_GTEST -DUSE_RPC -DUSE_TENSORPIPE -I/opt/pytorch/Fuser -I/opt/pytorch/Fuser/csrc -I/opt/pytorch/Fuser/lib/dynamic_type/src -I/opt/pytorch/Fuser/third_party/flatbuffe
rs/include -isystem /opt/pytorch/Fuser/third_party/googletest/googletest/include -isystem /opt/pytorch/Fuser/third_party/googletest/googlemock/include -isystem /opt/pytorch/Fuser/third_party/googletest/googletest -isystem /opt/pytorch/Fuser/third_party/googletest/googlemock -isystem /usr/local/lib/python3.10/dist-packages/torch/include -isystem /usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/api/include -isystem /usr/local/cuda/include -Wno-psabi -D_GLIBCXX_USE_CXX11_ABI=1 -g -Wall -Wno-unused-function -Werror -D_GLIBCXX_USE_CXX11_ABI=1 -MD -MT CMakeFiles/nvfuser_tests.dir/test/test_gpu_compute_with.cpp.o -MF CMakeFiles/nvfuser_tests.dir/test/test_gpu_compute_with.cpp.o.d -o CMakeFiles/nvfuser_tests.dir/test/test_gpu_compute_with.cpp.o -c /opt/pytorch/Fuser/test/test_gpu_compute_with.cpp                                        In file included from /opt/pytorch/Fuser/test/test_gpu_compute_with.cpp:37:
/opt/pytorch/Fuser/test/utils.h: In function ‘bool nvfuser::maybeClearAllocator(int64_t)’:
/opt/pytorch/Fuser/test/utils.h:400:26: error: cannot convert ‘c10::DeviceIndex*’ {aka ‘signed char*’} to ‘int*’
  400 |     c10::cuda::GetDevice(&device_index);
```
inside my docker that has Pytorch version 2.3.0a0+git38c18f3

Is it is not expected right?

**Comment by samnordmann** [comment]
> I have check out to this branch but am still getting the error
> 
> ```
> FAILED: CMakeFiles/nvfuser_tests.dir/test/test_gpu_compute_with.cpp.o
> /usr/bin/c++ -DNVFUSER_DISTRIBUTED -DUSE_C10D_GLOO -DUSE_C10D_MPI -DUSE_C10D_NCCL -DUSE_C10D_UCC -DUSE_DISTRIBUTED -DUSE_GTEST -DUSE_RPC -DUSE_TENSORPIPE -I/opt/pytorch/Fuser -I/opt/pytorch/Fuser/csrc -I/opt/pytorch/Fuser/lib/dynamic_type/src -I/opt/pytorch/Fuser/third_party/flatbuffe
> rs/include -isystem /opt/pytorch/Fuser/third_party/googletest/googletest/include -isystem /opt/pytorch/Fuser/third_party/googletest/googlemock/include -isystem /opt/pytorch/Fuser/third_party/googletest/googletest -isystem /opt/pytorch/Fuser/third_party/googletest/googlemock -isystem /usr/local/lib/python3.10/dist-packages/torch/include -isystem /usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/api/include -isystem /usr/local/cuda/include -Wno-psabi -D_GLIBCXX_USE_CXX11_ABI=1 -g -Wall -Wno-unused-function -Werror -D_GLIBCXX_USE_CXX11_ABI=1 -MD -MT CMakeFiles/nvfuser_tests.dir/test/test_gpu_compute_with.cpp.o -MF CMakeFiles/nvfuser_tests.dir/test/test_gpu_compute_with.cpp.o.d -o CMakeFiles/nvfuser_tests.dir/test/test_gpu_compute_with.cpp.o -c /opt/pytorch/Fuser/test/test_gpu_compute_with.cpp                                        In file included from /opt/pytorch/Fuser/test/test_gpu_compute_with.cpp:37:
> /opt/pytorch/Fuser/test/utils.h: In function ‘bool nvfuser::maybeClearAllocator(int64_t)’:
> /opt/pytorch/Fuser/test/utils.h:400:26: error: cannot convert ‘c10::DeviceIndex*’ {aka ‘signed char*’} to ‘int*’
>   400 |     c10::cuda::GetDevice(&device_index);
> ```
> 
> inside my docker that has Pytorch version 2.3.0a0+git38c18f3
> 
> Is it is not expected right?

Ok I understand, I guess I shouldn't use this pytorch version anymore, but either a release version or updated nightly version

**Comment by jjsjann123** [comment]
> There seems to be a build failure on distributed module happening at the same time. 😟
> 
> [pytorch/pytorch@f3e7d80#diff-eaf1a15a7c5a9d26268e146ab6203665f2c32686fef2b4308879b6f1413e476eR446-R447](https://github.com/pytorch/pytorch/commit/f3e7d809936d9f1bf63102e8afe241e13ed8766a#diff-eaf1a15a7c5a9d26268e146ab6203665f2c32686fef2b4308879b6f1413e476eR446-R447)
> 
> [pytorch/pytorch#119421](https://github.com/pytorch/pytorch/pull/119421)

strangely that issue didn't show up in my local build. Let me try again....

**Comment by xwang233** [comment]
https://github.com/pytorch/pytorch/pull/119421 This was reverted yesterday and seems like upstream is reland-ing it now.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
I'm just going to wait until the second CI pass before merging this one with admin hammer.

**Comment by jjsjann123** [comment]
landing as-is, since build finishes.

The nccl API change I'll try to patch in a different PR.

**Comment by jjsjann123** [comment]
Admin hammered it in.

Failing CI are from the re-landing of the other nccl API change. I'll patch that.

**Pull request merged**
